### DISCUSSION TITLE
All GitHub pages hang in GNOME Web

### DISCUSSION THREAD

**Discussion by xry111** [community/community#49755] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

Today all GitHub pages suddenly start to hang when I open them in GNOME Web (or Epiphany, a WebKitGTK based browser).  The console keeps outputting:

[Error] Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'navigator.clipboard.read')
	K (vendors-node_modules_stacktrace-parser_dist_stack-trace-parser_esm_js-node_modules_github_bro-327bbf-0aaeb22dd2a5.js:1:8117)
	every
	(anonymous function) (environment-7dd847af758e.js:1:5328)
	h (environment-7dd847af758e.js:1:5291)
	(anonymous function) (environment-7dd847af758e.js:1:4196)
	(anonymous function)

It looks like the JavaScript code is trapped in a dead loop.  I can pause all script execution via the Inspector of WebKit so I can at least make the report.

If you think it's a WebKitGTK bug I can report it to WebKit bugzilla.

**Comment by karlcow** [comment]
Reported on https://bugs.webkit.org/show_bug.cgi?id=253793
and https://github.com/webcompat/web-bugs/issues/119474

**Comment by karlcow** [comment]
https://bugs.webkit.org/show_bug.cgi?id=253758

**Comment by 4nd3r** [comment]
> If you think it's a WebKitGTK bug I can report it to WebKit bugzilla.

This is a GitHub bug, because they made assumption that every browser under the sun support clipboard API and/or failed to properly test it. That has negative impact on trusting GitHub. Whoever approved this change, I hope they will do proper postmortem.

**Reply by janxkoci** [reply]
Wasn't Github acquired by Microsoft? Sounds about right... :roll_eyes:

**Reply by sideeffect42** [reply]
Yes, Microsoft owns GitHub.

**Comment by sideeffect42** [comment]
> If you think it's a WebKitGTK bug I can report it to WebKit bugzilla.

I agree with @4nd3r on this. Even if it was a WebKit bug, GitHub should still work without having clipboard access.
I'm not even sure if I want my browser to expose access to the system clipboard to random websites...

If such a small error can cause the whole of GitHub to become completely unusable this shows to me that GitHub also has bigger issues in their error handling code which need some attention.

My takeaway from this is that I certainly need to diversify my code hosting strategy.

**Comment by ernstki** [comment]
It _sounds_ like a patch landed and was backported to an upcoming (?) libwebkit2gtk 2.40 release[^fn0], however I don't have the wherewithal to verify that at the moment, so consider it third-hand information.

This isn't a great consolation to those of us on some Ubuntu-20.04-derived distro[^fn1], who will probably be stuck at 2.38.something for the next few years. I'll confirm in a minute here, but I suspect the latest version of GNOME Web on Flathub (as of this writing, 44.1), with up-to-date runtimes, no longer suffers from this issue.

If you're stuck in some other situation, try this workaround[^fn2]

```javascript
navigator.clipboard = {};
```

in your `user-javascript.js`, which can be accessed under _Appearance_ → _Use Custom JavaScript_ in the GNOME Web settings.

[^fn0]: https://github.com/webcompat/web-bugs/issues/119474#issuecomment-1472757271
[^fn1]: I'm using elementaryOS 6.1 (Jólnir), which has their own Flatpak repo and their own release cycle for "sanctioned" Flatpak apps.
[^fn2]: https://bugs.webkit.org/show_bug.cgi?id=253758#c3

**Reply by ernstki** [reply]
I don't really know how to interpret these version numbers (other than later ones are bigger), but can confirm that the Flatpak for GNOME Web 44.1 doesn't seem to have this issue.

Or, rather, that runtime/org.gnome.Platform/x86_64/44 doesn't seem to have the issue.

```console
me@host:~/.local/share/flatpak/runtime/org.gnome.Platform/x86_64/43/active/files/lib/x86_64-linux-gnu$ ls -l libwebkit*
lrwxrwxrwx 1 me me       26 Jan 12 21:53 libwebkit2gtk-4.1.so.0 -> libwebkit2gtk-4.1.so.0.2.6*
-rwxr-xr-x 2 me me 66731352 Dec 31  1969 libwebkit2gtk-4.1.so.0.2.6*
lrwxrwxrwx 1 me me       26 Jan 12 21:53 libwebkit2gtk-5.0.so.0 -> libwebkit2gtk-5.0.so.0.0.0*
-rwxr-xr-x 2 me me 66741112 Dec 31  1969 libwebkit2gtk-5.0.so.0.0.0*

me@host:~/.local/share/flatpak/runtime/org.gnome.Platform/x86_64/44/active/files/lib/x86_64-linux-gnu$ ls -l libwebkit*
lrwxrwxrwx 1 me me       26 Apr  4 23:12 libwebkit2gtk-4.1.so.0 -> libwebkit2gtk-4.1.so.0.8.1*
-rwxr-xr-x 2 me me 74872264 Dec 31  1969 libwebkit2gtk-4.1.so.0.8.1*
lrwxrwxrwx 1 me me       25 Apr  4 23:12 libwebkitgtk-6.0.so.4 -> libwebkitgtk-6.0.so.4.0.1*
-rwxr-xr-x 2 me me 73703120 Dec 31  1969 libwebkitgtk-6.0.so.4.0.1*
```

**Reply by aperezdc** [reply]
> It _sounds_ like a patch landed and was backported to an upcoming (?) libwebkit2gtk 2.40 release[1](#user-content-fn-fn0-5f8195ddbdc57d1cceac47475571b1cc), however I don't have the wherewithal to verify that at the moment, so consider it third-hand information.

Indeed, the patch was backported to the 2.40 release branch, and version GitHub works as expected with 2.40.0.

> This isn't a great consolation to those of us on some Ubuntu-20.04-derived distro[2](#user-content-fn-fn1-5f8195ddbdc57d1cceac47475571b1cc), who will probably be stuck at 2.38.something for the next few years. I'll confirm in a minute here, but I suspect the latest version of GNOME Web on Flathub (as of this writing, 44.1), with up-to-date runtimes, no longer suffers from this issue.

I do think it will be useful for some folks to have one more 2.38.x release, so I have also backported the patch to the 2.38 release branch, and I am planning to cut one final release from the series, which will be 2.38.6&mdash;that version will include the fix, but it may still take a week or two because first we need to get a 2.40.1 out of the door, which has a bit more of priority.

Nevertheless, I also agree that the GitHub folks have put out an update which is buggy, and fails to properly do feature detection of available Web APIs. I want to think it was caused by unintentional lack of testing, but there has been plenty of time since the issue started happening to roll out a fix&mdash;which has not happened&mdash;and that is appalling.

**Reply by ernstki** [reply]
Thanks for your efforts, @aperezdc.

**Reply by aperezdc** [reply]
Both the 2.38.6 and 2.40.1 releases have been published last week, and they include the async clipboard API patch that makes GitHub work.
